README.md
## ipy2d

[![ipy2d on pypi](https://badgen.net/pypi/v/ipy2d)](https://pypi.org/project/ipy2d/)

A small library to convert IP addresses to integers

### Usage

```py
from ipy2d import convert

print(convert.from_4("127.0.0.1")) # -> 2130706433
print(convert.to_4(134744072)) # -> 8.8.8.8

print(convert.from_6("::1")) # -> 1
print(convert.to_6(1051570404137199630024704)) # -> 0000:0000:0000:dead:beef:0000:0000:0000
print(convert.to_6(1051570404137199630024704, compressed=True)) # -> ::dead:beef:0:0:0
```


CHANGELOG.md
# v0.0.7

## Features

* Added CHANGELOG.md and devlopment.md

# v0.0.6

## Features

* Improved README file

# v0.0.5

## Features

* Now you can compress IPv6 addresses

```py
from ipy2d import convert

print(convert.to_6(54321, compressed=True))
```


test/test.py
from ipy2d import fun

def main():
    assert fun.from_4("0.0.0.1") == 1, "Should be 1"
    assert fun.from_6("::1") == 1, "Should be 1"
    assert fun.from_6("2001:4860:4860::8888") == 42541956123769884636017138956568135816, "Should be 42541956123769884636017138956568135816"

    assert fun.to_4(5) == "0.0.0.5", "Should be 0.0.0.5"
    assert fun.to_6(42541956123769884636017138956568135817) == "2001:4860:4860:0000:0000:0000:0000:8889", "Should be 2001:4860:4860:0000:0000:0000:0000:8889"

main()


test/test_ipy2d.py
from ipy2d import convert
import pytest


def test_random_examples():
    assert convert.from_4("0.0.0.1") == 1, "Should be 1"
    assert convert.from_6("::1") == 1, "Should be 1"
    assert (
        convert.from_6("2001:4860:4860::8888") == 42541956123769884636017138956568135816
    ), "Should be 42541956123769884636017138956568135816"

    assert convert.to_4(5) == "0.0.0.5", "Should be 0.0.0.5"
    assert (
        convert.to_6(42541956123769884636017138956568135817)
        == "2001:4860:4860:0000:0000:0000:0000:8889"
    ), "Should be 2001:4860:4860:0000:0000:0000:0000:8889"
    assert (
        convert.to_6(0x20014860486000000000000000008889)
        == "2001:4860:4860:0000:0000:0000:0000:8889"
    ), "Should be 2001:4860:4860:0000:0000:0000:0000:8889"
    assert (
        convert.to_6(convert.from_6("0000:0000:0000:dead:beef:0000:0000:0000"))
        == "0000:0000:0000:dead:beef:0000:0000:0000"
    ), "Should be 0000:0000:0000:dead:beef:0000:0000:0000"


def test_first_two_octets_ipv4():
    ints = [i for i in range(0, 0x100)]
    for i1 in [0]:
        for i2 in [0]:
            for i3 in ints:
                for i4 in ints:
                    i = i1 << 0x18 | i2 << 0x10 | i3 << 0x08 | i4
                    assert (
                        convert.from_4(".".join(map(str, [i1, i2, i3, i4]))) == i
                    ), "Should be {0}".format(i)
                    assert convert.to_4(i) == ".".join(
                        map(str, [i1, i2, i3, i4])
                    ), ".".join(map(str, [i1, i2, i3, i4]))


def test_compressed_ipv6():
    assert "::1" == convert.to_6(1, compressed=True), "Should be ::1"
    assert "2001:4860:4860::8889" == convert.to_6(
        0x20014860486000000000000000008889, compressed=True
    ), "Should be 2001:4860:4860::8889"
    assert "1023::3a:3:0" == convert.to_6(
        0x1023_0000_0000_0000_0000_003A_0003_0000, compressed=True
    ), "Should be 1023::3a:3:0"
    assert "::dead:beef:0:0:0" == convert.to_6(
        1051570404137199630024704, compressed=True
    ), "Should be ::dead:beef:0:0:0"


@pytest.mark.skip(reason="this test takes to long!")
def test_all_ipv4():
    ints = [i for i in range(0x00, 0x100)]
    for i1 in ints:
        for i2 in ints:
            for i3 in ints:
                for i4 in ints:
                    i = i1 << 0x18 | i2 << 0x10 | i3 << 0x08 | i4
                    assert (
                        convert.from_4(".".join(map(str, [i1, i2, i3, i4]))) == i
                    ), "Should be {0}".format(i)
                    assert convert.to_4(i) == ".".join(
                        map(str, [i1, i2, i3, i4])
                    ), "Should be {0}".format(".".join(map(str, [i1, i2, i3, i4])))


ipy2d/convert.py
import netaddr
import ipaddress


def from_4(ipv4):
    octets = ipv4.split(".")
    return (
        int(octets[0]) << 0x18
        | int(octets[1]) << 0x10
        | int(octets[2]) << 0x08
        | int(octets[3]) << 0x00
    )


def from_6(ipv6):
    ip = netaddr.IPAddress(ipv6)
    hextet = ip.format(dialect=netaddr.ipv6_verbose).split(":")
    return (
        int(hextet[0], 0x10) << 0x70
        | int(hextet[1], 0x10) << 0x60
        | int(hextet[2], 0x10) << 0x50
        | int(hextet[3], 0x10) << 0x40
        | int(hextet[4], 0x10) << 0x30
        | int(hextet[5], 0x10) << 0x20
        | int(hextet[6], 0x10) << 0x10
        | int(hextet[7], 0x10) << 0x00
    )


def to_4(num):
    return "{0}.{1}.{2}.{3}".format(
        num >> 0x18 & 0xFF, num >> 0x10 & 0xFF, num >> 0x08 & 0xFF, num >> 0x00 & 0xFF
    )


def to_6(num, compressed=False):
    ipv6 = "{0:>04x}:{1:>04x}:{2:>04x}:{3:>04x}:{4:>04x}:{5:>04x}:{6:>04x}:{7:>04x}".format(
        num >> 0x70 & 0xFFFF,
        num >> 0x60 & 0xFFFF,
        num >> 0x50 & 0xFFFF,
        num >> 0x40 & 0xFFFF,
        num >> 0x30 & 0xFFFF,
        num >> 0x20 & 0xFFFF,
        num >> 0x10 & 0xFFFF,
        num >> 0x00 & 0xFFFF,
    )
    return ipv6 if not compressed else str(ipaddress.IPv6Address(ipv6).compressed)


ipy2d/fun.py
import netaddr
import ipaddress

def from_4(ipv4):
    octets = ipv4.split('.')
    return int(octets[0]) << 0x18 | \
           int(octets[1]) << 0x10 | \
           int(octets[2]) << 0x08 | \
           int(octets[3]) << 0x00

def from_6(ipv6):
    ip = netaddr.IPAddress(ipv6)
    hextet = ip.format(dialect=netaddr.ipv6_verbose).split(':')
    return int(hextet[0], 0x10) << 0x70 | \
           int(hextet[1], 0x10) << 0x60 | \
           int(hextet[2], 0x10) << 0x50 | \
           int(hextet[3], 0x10) << 0x40 | \
           int(hextet[4], 0x10) << 0x30 | \
           int(hextet[5], 0x10) << 0x20 | \
           int(hextet[6], 0x10) << 0x10 | \
           int(hextet[7], 0x10) << 0x00 

def to_4(num):
    return "{0}.{1}.{2}.{3}".format(num >> 0x18 & 0xff, \
                                    num >> 0x10 & 0xff, \
                                    num >> 0x08 & 0xff, \
                                    num >> 0x00 & 0xff)
                                
def to_6(num, compressed=False):
    ipv6 = "{0:>04x}:{1:>04x}:{2:>04x}:{3:>04x}:{4:>04x}:{5:>04x}:{6:>04x}:{7:>04x}".format( \
        num >> 0x70 & 0xffff, \
        num >> 0x60 & 0xffff, \
        num >> 0x50 & 0xffff, \
        num >> 0x40 & 0xffff, \
        num >> 0x30 & 0xffff, \
        num >> 0x20 & 0xffff, \
        num >> 0x10 & 0xffff, \
        num >> 0x00 & 0xffff)
    return ipv6 if not compressed else str(ipaddress.IPv6Address(ipv6).compressed)